Organ movement refactor *Un-nullspaces your organs*

Time-Green opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

closes #53931, #70916, #53931

## About The Pull Request

Organs were previously stored in nullspace. Now they are stored in their prospective bodyparts. Bodyparts are now stored in the mob.

I've also had to refactor a lot of code concerning organ movement. Previously, organs were only moved into bodyparts once the bodyparts were removed. To accomodate this change, two major distinctions have been made:

**Bodypart removal/insertion**
Called only when an organ is taken out of a bodypart. Bodypart overlays, damage modifiers or other changes that should affect a bodypart itself goes here. 

**Mob insertion/removal**
Called when an organ is removed from a mob. This can either be directly, by taking the organ out of a mob, or by removing the bodypart that contains the organ. This lets you add and remove organ effects safely without having to worry about the bodypart. 

Now that we controle the movement of bodyparts and organs, we can fuck around with them more. Summoning someones head or chest or heart will actually kill them now (and quite violently I must say (chest summoning gibs lol)). 

https://github.com/tgstation/tgstation/assets/7501474/5efc9dd3-cfd5-4ce4-b70f-d0d74894626e

I´ve also added a unit test that violently tears apart and reconstructs a person in different ways to see if they get put toghether the right way

This will definitely need a testmerge. I've done a lot of testing to make sure interactions work, but more niche stuff or my own incompetence can always slip through. 

## Why It's Good For The Game

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

A lot of organ work is quite restricted. You can't C4 someones heart, you cant summon their organs and a lot of exceptions have to be made to keep organs in nullspace. This lets organs (and bodyparts) play more nicely with the rest of the game. This also makes it a lot easier to move away from extorgans since a lot of their unique movement code has been removed and or generalized. 

I don't like making PRs of this size (I'm so sorry reviewers), but I was in a unique position to replace the entire system in a way I couldn't have done conveniently in multiple PRs

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
refactor: Your organs are now inside your body. Please report any issues with bodypart and organ movement, including exotic organ, on github and scream at me
fix: Cases of unexpected organ movement, such as teleporting bodyparts and organs with spells, now invokes a proper reaction (usually violent death)
runtime: Fixes HARS runtiming on activation/deactivation
fix: Fixes lag when species swapping
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Time-Green committed:
Tracks organs on limbs

Time-Green committed:
Puts organs into limb contents

Time-Green committed:
Reorganizes and deletes copy paste, adds safeties

Time-Green committed:
Cleans up set_owner(), removes deprecated code

Time-Green committed:
Cleans up leg code, unmixes your params

Time-Green committed:
I know what needs to be done ;_;

Time-Green committed:
seperates mob and limb organ removal

Time-Green committed:
Unit tests and final adjustments

Time-Green committed:
Merge branch 'master' into moves-organs-inside-of-limbs

Time-Green committed:
post-merge

Time-Green committed:
Kills legacy organ movement procs

Time-Green committed:
limb to bodypart

Time-Green committed:
sanity, clean-up and chiggles your brainmob

Time-Green committed:
Fixes runtime in head unorganing

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into moves-organs-inside-of-limbs

SyncIt21 commented:
- Fixes #53931

MrMelbert commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -201,48 +185,56 @@
+		head.brainmob = null
+		brainmob.forceMove(src)
+
+/obj/item/organ/internal/eyes/on_bodypart_insert(obj/item/bodypart/head/head)
+	. = ..()
 	head.eyes = src
 	..()
```
Double parent call

MrMelbert commented on code/modules/surgery/bodyparts/species_parts/misc_bodyparts.dm:
```diff
@@ -512,7 +512,7 @@
 	return ..()
 
-/obj/item/bodypart/arm/left/golem/set_owner(new_owner)
+/obj/item/bodypart/arm/left/golem/update_owner(new_owner)
```
Why is this not `apply_ownership` and `clear_ownership`?

Time-Green replied:
I don't know tbh

MrMelbert commented on code/modules/surgery/bodyparts/species_parts/misc_bodyparts.dm:
```diff
@@ -546,7 +546,7 @@
 	return ..()
 
-/obj/item/bodypart/arm/right/golem/set_owner(new_owner)
+/obj/item/bodypart/arm/right/golem/update_owner(new_owner)
```
Ditto

MrMelbert commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -736,70 +726,79 @@
-///Proc to change the value of the `owner` variable and react to the event of its change.
-/obj/item/bodypart/proc/set_owner(new_owner)
+/// Proc to change the value of the `owner` variable and react to the event of its change.
+/obj/item/bodypart/proc/update_owner(new_owner)
```
At a glance it looks like you want to make this `SHOULD_NOT_OVERRIDE` rather than `SHOULD_CALL_PARENT`

MrMelbert reviewed (changes requested):
Getting wabbajacked instantly gibs you, because it forces removal of your chest bodypart. Likely the same for a few other transformations (`Animalize` for example, or going through the borg machine)

MrMelbert commented:
Not sure if you care, but `export_item_and_contents` will sell all of a mob's organs out of their body. This is particularly noticeable with the pirate's cargo hold pad (if a pirate is standing on the pad when it ships, not only will it sell all of their equipment but it will ship all of their vital organs and KO them.)

Kapu1178 commented:
I did this on Daedalus, it creates some funny bugs, here's a couple fixes for you:
1. Heretic sacrifice thingy checks `range()` instead of `orange()`, so this includes your own organs and bodyparts
2. Crafting does the same.

Time-Green commented:
> I did this on Daedalus, it creates some funny bugs, here's a couple fixes for you:
> 
> 1. Heretic sacrifice thingy checks `range()` instead of `orange()`, so this includes your own organs and bodyparts
> 2. Crafting does the same.

Fucking nice! I’ll put them on my todo

@MrMelbert also good callouts. I’ll get on fixing them tomorrow. I indeed did not check those transforms

CCC23 commented:
new funny gimmic mark someones augment chest for garyock implant and use power word testicular torsion on a man

Kapu1178 commented:
> > I did this on Daedalus, it creates some funny bugs, here's a couple fixes for you:
> > 
> > 1. Heretic sacrifice thingy checks `range()` instead of `orange()`, so this includes your own organs and bodyparts
> > 2. Crafting does the same.
> 
> Fucking nice! I’ll put them on my todo
> 
> @MrMelbert also good callouts. I’ll get on fixing them tomorrow. I indeed did not check those transforms
> 
> Selling someones organs while inside someone also seems unintentional, I’ll get on it

do what i did and make organs inside of people have the ABSTRACT item flag, and change relevant loops to exclude ABSTRACT items. fixes other bugs too.

ChungusGamer666 commented:
Oh look it's the stuff I was working on and had basically entirely done until Oranges had a shitfit and banned me for 30 days. Wholesome!

ChungusGamer666 commented:
> I did this on Daedalus, it creates some funny bugs, here's a couple fixes for you:
> 
>     1. Heretic sacrifice thingy checks `range()` instead of `orange()`, so this includes your own organs and bodyparts
> 
>     2. Crafting does the same.

Bodyparts are still stored in nullspace, so I don't see how this could happen...

Jacquerel commented:
> Oh look it's the stuff I was working on and had basically entirely done until Oranges had a shitfit and banned me for 30 days. Wholesome!

You could have PRed it after your ban. Next time I would advise not getting banned.

ChungusGamer666 commented:
> You could have PRed it after your ban. Next time I would advise not getting banned.

Nah, I'd rather PR it literally anywhere else.

necromanceranne commented:
Have you made sure explosions don't delete organs inside mobs?

Time-Green commented:
> Have you made sure explosions don't delete organs inside mobs?

I’m pretty sure I prevented them from propogating into someones insides, I’ll double check

Time-Green committed:
Cleans up qdel ordering

Time-Green committed:
Protects contents

Time-Green committed:
Cleans up some update_ownership code

Time-Green commented:
Fixed contents explosion, transform bugs and cleaned up the update ownership stuff.
I could not actually replicate the heretic and bounty bugs. They did not reach organs or limbs, so I try not to think too hard about it

Time-Green committed:
Qdel fixes, bodypart cleaning and abstract flags p1

Time-Green committed:
FUCK BRAINMOB

Time-Green committed:
Adds organ movement flags

Time-Green committed:
Unit test error

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into moves-organs-inside-of-limbs

Time-Green committed:
linter moment

ZephyrTFA commented on code/modules/antagonists/nightmare/nightmare_organs.dm:
```diff
@@ -27,9 +27,10 @@
 		terrorize_spell.Grant(brain_owner)
 
+/obj/item/organ/internal/brain/shadow/nightmare/on_mob_remove(mob/living/carbon/brain_owner)
+
 	RegisterSignal(brain_owner, COMSIG_ATOM_PRE_BULLET_ACT, PROC_REF(dodge_bullets))
 
-/obj/item/organ/internal/brain/shadow/nightmare/on_remove(mob/living/carbon/brain_owner)
 	. = ..()
```
this intentional?

Time-Green replied:
It is not

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into moves-organs-inside-of-limbs

Time-Green committed:
Optimizes mindmovement

Time-Green committed:
Clean up, fixes a mistake I made

Time-Green committed:
scope fix

Time-Green committed:
Fixes species change

Time-Green committed:
huh

Time-Green committed:
Makes the unit test checkmark a green checkmark

Time-Green committed:
just kidding this solves the unit tests

Time-Green committed:
Fixes monkeys grabbing bodyparts and internal organs from you

Time-Green committed:
Rare QDEL race condition I am going insane

Time-Green committed:
body

Time-Green committed:
I'm so tired boss

Time-Green committed:
Removes useless nullchecks, adds useful check

Time-Green committed:
Fixes harddels (I prayeth)

Time-Green committed:
comitting unit test doesnt count

Time-Green committed:
this test did not make sense

Time-Green committed:
where am i

Time-Green committed:
I am happy and healthy

Time-Green committed:
I am damaged

Time-Green commented:
Ignore commit count, unit tests take too long to run on my PC and I've been using github as long term testing.
This PR will need a testmerge

Time-Green commented:
This does not play very well with polymorph and changelings... back to the drawing board

Time-Green committed:
Fixes labor shuttle deleting your organs

And changelings breaking

Kapu1178 commented on code/modules/admin/verbs/admin.dm:
```diff
@@ -176,6 +176,10 @@
 		if(!M.dropItemToGround(W))
+			// I hate that this is necessary, but the code is literally just dropping or deleting everything otherwise
+			// people should be allowed to keep their fucking organs
+			if(istype(W, /obj/item/organ) || istype(W, /obj/item/bodypart))
```
item abstract tho?

Time-Green replied:
It’s an admin verb to drop or delete everything, abstract flags are also given to things like armblades which an admin would want to be gone

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into moves-organs-inside-of-limbs

Time-Green committed:
Now actually stops labor console from deconstructing people

GoblinBackwards commented:
Experienced 2 bugs last time this was testmerged that I'm pretty confident were caused by this pr.

1: When I was shot with a bolt of change that transformed me into another human I was kicked out of my body, with my body becoming catatonic.

2: When I crafted a head on a spear using someone's decapitated head off the floor, it took my head off my body and put that on the spear instead. The head on the floor that should have been used seemed to also get deleted when this happened.

timothymtorres commented:
Someone attempted to do this already in #53934

This also might fix:
- #52103
- #70916
- #53931

Time-Green committed:
Crafting respects abstract, logs forced gibbing

AdipemDragon commented:
A TM thing I've seen experienced twice already, which. I am unsure if it's a bug or not.... is how a brain will continue to ammass brain damage for no discernible reason after a head is transplanted onto another body, for the sake of a quick fix

Time-Green committed:
fix living decay after organ insertion

Time-Green commented:
> A TM thing I've seen experienced twice already, which. I am unsure if it's a bug or not.... is how a brain will continue to ammass brain damage for no discernible reason after a head is transplanted onto another body, for the sake of a quick fix

Decaying while alive is sadly not a feature. Fixed in newest version

ATH1909 commented:
I lowkey would prefer it if summoning an organ out of someone's body didn't give them and instead let them die naturally (due to the consequences of lacking that organ), but maybe that's just a personal preference. IMO, ripping someone's head off and watching their now-headless body slump over is MUCH more satisfying than a generic gib animation.

Jezzared commented:
> I lowkey would prefer it if summoning an organ out of someone's body didn't gib them and instead let them die naturally (due to the consequences of lacking that organ), but maybe that's just a personal preference. IMO, ripping someone's head off and watching their now-headless body slump over is MUCH more satisfying than a generic gib animation.

I have to agree. Also includes it in the sandbox more, removing an organ from a creature that doesn't care about that organ probably shouldn't case it to explode.

"Local doctor saves man by summoning his inflamed appendix with magic."

Time-Green commented:
> I lowkey would prefer it if summoning an organ out of someone's body didn't gib them and instead let them die naturally (due to the consequences of lacking that organ), but maybe that's just a personal preference. IMO, ripping someone's head off and watching their now-headless body slump over is MUCH more satisfying than a generic gib animation.

Is this something you experienced? Summoning someones head doesn’t gib, it just tears their head off. Gibbing occurs when a mob loses their chest

AdipemDragon commented:
> Did you run into any other bugs?

Well... 
![image](https://github.com/tgstation/tgstation/assets/76874615/8466e946-68d9-4e23-8240-d20465e1e9bf)
Plasmanification seems fucked, as as soon as the chest gets plasmamanned, uhh... That

ATH1909 commented:
> > I lowkey would prefer it if summoning an organ out of someone's body didn't gib them and instead let them die naturally (due to the consequences of lacking that organ), but maybe that's just a personal preference. IMO, ripping someone's head off and watching their now-headless body slump over is MUCH more satisfying than a generic gib animation.
> 
> Is this something you experienced? Summoning someones head doesn’t gib, it just tears their head off. Gibbing occurs when a mob loses their chest

oh, the opening post implied that it happens when you rip out someone's heart. if it only happens when you rip out the chest, that's fine.

Time-Green commented:
> > > I lowkey would prefer it if summoning an organ out of someone's body didn't gib them and instead let them die naturally (due to the consequences of lacking that organ), but maybe that's just a personal preference. IMO, ripping someone's head off and watching their now-headless body slump over is MUCH more satisfying than a generic gib animation.
> > 
> > 
> > Is this something you experienced? Summoning someones head doesn’t gib, it just tears their head off. Gibbing occurs when a mob loses their chest
> 
> oh, the opening post implied that it happens when you rip out someone's heart. if it only happens when you rip out the chest, that's fine.

Well, that’s not completely wrong. Gar’yok grabs the top container



> > Did you run into any other bugs?
> 
> Well... ![image](https://private-user-images.githubusercontent.com/76874615/289080268-8466e946-68d9-4e23-8240-d20465e1e9bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwNTk1MzgsIm5iZiI6MTcwMjA1OTIzOCwicGF0aCI6Ii83Njg3NDYxNS8yODkwODAyNjgtODQ2NmU5NDYtNjhkOS00ZTIzLTgyNDAtZDIwNDY1ZTFlOWJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA4VDE4MTM1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjYzJmYWNkZDRkYWJjZTA4ZWM0YTA4ODdjMWQwN2ZmYTdjNTE0YThkZTM4YTA0NmM4MjMxZjRiNTdiZDNkOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7DUckmaPux3ORjfd2dpFH3ZjXzqNH7dkxudeoQT3OO0) Plasmanification seems fucked, as as soon as the chest gets plasmamanned, uhh... That

Paiiiiiin. Plasmalimb replacement is probably just dumb, this doesn’t usually happen

ATH1909 commented:
> Well, that’s not completely wrong. Gar’yok grabs the top container

awwwwwwww

Isratosh commented:
Wizard's wand of polymorph gibs humans with this admin log
ADMIN LOG: *no key*/(Alessandra Taylor) was gibbed after their chest teleport to Thunderdome Arena (171,52,1)
Round ID 220190 with a real wizard and round ID 220193 with me testing

AdipemDragon commented:
> Paiiiiiin. Plasmalimb replacement is probably just dumb, this doesn’t usually happen

could my spaceman having a prosthetic limb in the moment have influenced that in anyway?

Time-Green commented:
> Wizard's wand of polymorph gibs humans with this admin log
> ADMIN LOG: _no key_/(Alessandra Taylor) was gibbed after their chest teleport to Thunderdome Arena (171,52,1)
> Round ID 220190 with a real wizard and round ID 220193 with me testing

That makes so little sense, that worked perfectly? Aargh



> > Paiiiiiin. Plasmalimb replacement is probably just dumb, this doesn’t usually happen
> 
> could my spaceman having a prosthetic limb in the moment have influenced that in anyway?

It shouldn’t’ve, limb replacement was fine in my testing so it’s probably specific to plasma rivers

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into moves-organs-inside-of-limbs

Time-Green committed:
Fixes limb replacement and polymorph gibbing

Kapu1178 commented on code/modules/surgery/bodyparts/helpers.dm:
```diff
@@ -15,6 +15,7 @@
 /mob/living/carbon/proc/del_and_replace_bodypart(obj/item/bodypart/new_limb, special)
 	var/obj/item/bodypart/old_limb = get_bodypart(new_limb.body_zone)
 	if(old_limb)
+		old_limb.drop_limb(special = TRUE)
 		qdel(old_limb)
```
this shouldn't be necessarily, qdeling a limb should implicitly specially remove it

Time-Green replied:
I blackmailed melon into merging it. I’ll look into it when I get to my computer, I forgot the finer details of drop_limb
